### PR TITLE
mz - adding "suspend" button to users table for admins

### PR DESCRIPTION
**Pull request by mczimmermann**
## Overview
In this PR, I added an "suspend" button column on the users table for admins to be able to remove certain users from ALL commons if they are causing issues. Currently, the button serves as a placeholder and clicking it does not actually redirect/do anything.

## Screenshots
Screen Shot 2024-05-27 at 5 09 36 PM

## Feedback Request 
Please give suggestions on the following:
- Should users that are admins not have "Suspend" buttons next to them? Would it ever be necessary to suspend an admin from a commons?
- I'm not sure if this is the most intuitive design, I'm open to other ideas about how it could be implemented better. 

## Future Possibilities 
Some possible points could be:
- This is the beginning/part of the "suspend", and we will need to implement the actual remove page for it to work.

## Validation (Optional)
1. Download this branch.
2. Run the project.
3. Go to "Users" under "Admin".
4. Check that for all users you can see a button that says "Suspend".

## Tests
- [x] Backend Unit tests (`mvn test`) pass
- [x] Backend Test coverage (`mvn test jacoco:report`) 100%
- [x] Backend Mutation tests (`mvn test pitest:mutationCoverage`) 100% 
- [x] Frontend Unit tests (`npm test`) pass
- [x] Frontend Test coverage (`npm run coverage`) 100%
- [x] Frontend Mutation tests (`npx stryker run`) 100% 
- [x] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues
Related to #8, does not close it. 
Addresses sub-issue F1 within [EPIC:Admin can suspend/restore Users](https://github.com/ucsb-cs156/proj-happycows/issues/124)

## Dokku Deployment
[https://happycows-mczimmermann-dev.dokku-07.cs.ucsb.edu](https://happycows-mczimmermann-dev.dokku-07.cs.ucsb.edu)

### CODE REVIEW

File: frontend/src/main/components/Users/UsersTable.js [hunk 1]
```diff
@@ -1,37 +1,50 @@
...
+            Header: 'Admin',
+            id: 'admin',
+            accessor: (row, _rowIndex) => String(row.admin) // hack needed for boolean values to show up
+        },
+    ];
+
+    // if (hasRole(currentUser, "ROLE_ADMIN")) {
+    // don't need to check for admin because this page will not display if not admin
+    columns.push(ButtonColumn("Edit Commons", "danger", removeCallback, "UsersTable"));
+    // }
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]: https://ucsb-cs156.github.io/topics/code_reviews/commented_out_code.htm

File: frontend/src/main/components/Users/UsersTable.js [hunk 2]
```diff
@@ -1,37 +1,50 @@
...
-        Header: 'Admin',
-        id: 'admin',
-        accessor: (row, _rowIndex) => String(row.admin) // hack needed for boolean values to show up
-    },
-];
 
 export default function UsersTable({ users }) {
+
+    // Stryker disable next-line all : TODO create this function & test it
+    const removeCallback = (_cell) => { }
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]: Why do we have a callback that does nothing?

**Review by pconrad — CHANGES_REQUESTED** [review]
> Currently, the button serves as a placeholder and clicking it does not actually redirect/do anything.

Typically, we try to not merge code into main that is broken.  It is true that we like to break PRs into small pieces, but in this case, it's "too small a piece".  Please at least link to a placeholder page for the edit page that says something like "Edit not implemented yet; coming soon".

Also see other comments in the review about "commented out code"

**Comment by mczimmermann** [comment]
Screen Shot 2024-05-21 at 10 53 11 PM
UPDATED view

**Review by nadavital — DISMISSED** [review]
LGTM

**Review by nadavital — DISMISSED** [review]
LGTM

**Review by AnikaArora — CHANGES_REQUESTED** [review]
Link to dokku deployment?

**Comment by mczimmermann** [comment]
> Link to dokku deployment?

Just added it to the PR!

File: frontend/src/App.js [hunk 3]
```diff
@@ -19,6 +19,8 @@ import { hasRole, useCurrentUser } from "main/utils/currentUser";
 import NotFoundPage from "main/pages/NotFoundPage";
 import AdminViewPlayPage from "main/pages/AdminViewPlayPage";
 import ProtectedPlayPage from "main/pages/ProtectedPlayPage";
+import AdminRemoveUserPage from "main/pages/AdminRemoveUserPage";
```
Comment (pconrad) [inline frontend/src/App.js]:
```suggestion
import AdminSuspendUserPage from "main/pages/AdminSuspendUserPage";
```

Let's change this to "suspend user", per: #30

File: frontend/src/App.js [hunk 4]
```diff
@@ -48,6 +50,10 @@ function App() {
                 path="/admin/play/:commonsId/user/:userId"
                 element={<AdminViewPlayPage />}
             />
+            <Route 
+                path="/admin/remove/user/:userId" 
+                element={<AdminRemoveUserPage />}
```
Comment (pconrad) [inline frontend/src/App.js]:
```suggestion
                element={<AdminSuspendUserPage />}
```

File: frontend/src/main/components/Users/UsersTable.js [hunk 5]
```diff
@@ -1,37 +1,49 @@
...
+        },
+        {
+            Header: 'Admin',
+            id: 'admin',
+            accessor: (row, _rowIndex) => String(row.admin) // hack needed for boolean values to show up
+        },
+    ];
+
+    // don't need to check for admin because this page will not display if not admin
+    columns.push(ButtonColumn("Edit Commons", "primary", removeCallback, "UsersTable"));
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]:
```suggestion
    columns.push(ButtonColumn("Suspend", "primary", suspendCallback, "UsersTable"));
```

File: frontend/src/main/components/Users/UsersTable.js [hunk 6]
```diff
@@ -1,37 +1,49 @@
...
-        accessor: (row, _rowIndex) => String(row.admin) // hack needed for boolean values to show up
-    },
-];
+import { useNavigate } from 'react-router-dom';
 
 export default function UsersTable({ users }) {
+
+    const navigate = useNavigate();
+
+    const removeCallback = (row) => {
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]:
```suggestion
    const suspendCallback = (row) => { 
```

File: frontend/src/main/components/Users/UsersTable.js [hunk 7]
```diff
@@ -1,37 +1,49 @@
...
-];
+import { useNavigate } from 'react-router-dom';
 
 export default function UsersTable({ users }) {
+
+    const navigate = useNavigate();
+
+    const removeCallback = (row) => { 
+        console.log(row)
+        navigate(`/admin/remove/user/${row.row.values.id}`)
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]:
```suggestion
        navigate(`/admin/suspend/user/${row.row.values.id}`)
```

File: frontend/src/main/pages/AdminRemoveUserPage.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import BasicLayout from "main/layouts/BasicLayout/BasicLayout";
+
+const AdminRemoveUserPage = () => {
```
Comment (pconrad) [inline frontend/src/main/pages/AdminRemoveUserPage.js]:
```suggestion
const AdminSuspendUserPage = () => {
```

File: frontend/src/main/pages/AdminRemoveUserPage.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+   // Stryker disable all : placeholder for future implementation
+  return (
+    <BasicLayout>
+      <div className="pt-2">
+        <h1>Edit User Commons page not yet implemented</h1>
```
Comment (pconrad) [inline frontend/src/main/pages/AdminRemoveUserPage.js]:
```suggestion
        <h1>Suspend User page not yet implemented</h1>
```

File: frontend/src/main/pages/AdminRemoveUserPage.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+      </div>
+    </BasicLayout>
+  )
+};
+
+export default AdminRemoveUserPage;
```
Comment (pconrad) [inline frontend/src/main/pages/AdminRemoveUserPage.js]:
```suggestion
export default AdminSuspendUserPage;
```

File: frontend/src/tests/components/Users/UsersTable.test.js [hunk 11]
```diff
@@ -1,24 +1,57 @@
...
     test("renders without crashing for three users", () => {
         render(
-            <UsersTable users={usersFixtures.threeUsers} />
+            <MemoryRouter>
+                <UsersTable users={usersFixtures.threeUsers} />
+            </MemoryRouter>
         );
     });
 
+    test("Remove button calls remove callback", async () => {
```
Comment (pconrad) [inline frontend/src/tests/components/Users/UsersTable.test.js]:
```suggestion
    test("Suspend button calls suspend callback", async () => {
```

File: frontend/src/tests/components/Users/UsersTable.test.js [hunk 12]
```diff
@@ -1,24 +1,57 @@
...
+        const testId = "UsersTable";
+        render(
+            <MemoryRouter>
+                <UsersTable users={usersFixtures.threeUsers} />
+            </MemoryRouter>
+        );
+
+        expect(screen.getByTestId(`${testId}-cell-row-0-col-id`)).toHaveTextContent("1");
+
+        const editButton = screen.getByTestId(`${testId}-cell-row-0-col-Edit Commons-button`);
```
Comment (pconrad) [inline frontend/src/tests/components/Users/UsersTable.test.js]:
```suggestion
        const editButton = screen.getByTestId(`${testId}-cell-row-0-col-Suspend-button`);
```

File: frontend/src/tests/components/Users/UsersTable.test.js [hunk 13]
```diff
@@ -1,24 +1,57 @@
...
+        expect(editButton).toBeInTheDocument();
+        expect(editButton).toHaveClass("btn-primary");
+        fireEvent.click(editButton);
+
+        await waitFor(() => expect(mockedNavigate).toHaveBeenCalledWith('/admin/remove/user/1'));
```
Comment (pconrad) [inline frontend/src/tests/components/Users/UsersTable.test.js]:
```suggestion
        await waitFor(() => expect(mockedNavigate).toHaveBeenCalledWith('/admin/suspend/user/1'));
```

File: frontend/src/tests/pages/AdminRemoveUserPage.test.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import { render, screen } from "@testing-library/react";
+import AdminRemoveUserPage from "main/pages/AdminRemoveUserPage";
```
Comment (pconrad) [inline frontend/src/tests/pages/AdminRemoveUserPage.test.js]:
```suggestion
import AdminSuspendUserPage from "main/pages/AdminSuspendUserPage";
```

File: frontend/src/tests/pages/AdminRemoveUserPage.test.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import { QueryClient, QueryClientProvider } from "react-query";
+import { MemoryRouter } from "react-router-dom";
+
+import { apiCurrentUserFixtures } from "fixtures/currentUserFixtures";
+import { systemInfoFixtures } from "fixtures/systemInfoFixtures";
+import axios from "axios";
+import AxiosMockAdapter from "axios-mock-adapter";
+
+
+describe("AdminRemoveUserPage tests", () => {
```
Comment (pconrad) [inline frontend/src/tests/pages/AdminRemoveUserPage.test.js]:
```suggestion
describe("AdminSuspendUserPage tests", () => {
```

File: frontend/src/tests/pages/AdminRemoveUserPage.test.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    test("Renders expected content", () => {
+        // arrange
+
+        setupAdminUser();
+
+        // act
+        render(
+            <QueryClientProvider client={queryClient}>
+                <MemoryRouter>
+                    <AdminRemoveUserPage />
```
Comment (pconrad) [inline frontend/src/tests/pages/AdminRemoveUserPage.test.js]:
```suggestion
                    <AdminSuspendUserPage />
```

File: frontend/src/tests/pages/AdminRemoveUserPage.test.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+                </MemoryRouter>
+            </QueryClientProvider>
+        );
+
+        // assert
+        expect(screen.getByText("Edit User Commons page not yet implemented")).toBeInTheDocument();
```
Comment (pconrad) [inline frontend/src/tests/pages/AdminRemoveUserPage.test.js]:
```suggestion
        expect(screen.getByText("Suspend User page not yet implemented")).toBeInTheDocument();
```

**Review by pconrad — CHANGES_REQUESTED** [review]
See discussion on the #proj-happycows Slack channel here:
* https://ucsb-cs156-s24.slack.com/archives/C073GT1NXUM/p1716580120753429

I think you can repurpose this PR as one that addresses part of issue #30 
and that way it's not at all wasted effort, even though we are going in a different direction.  

In my suggestions, note that I may have missed a few spots, and you will need to rename a file or two.    But you can implement that suggestions simply by clicking them one by one in the PR review (or you can just manually do them in your own editor; your choice.)

**Comment by mczimmermann** [comment]
Addresses sub-issue F1 within [EPIC:Admin can suspend/restore Users](https://github.com/ucsb-cs156/proj-happycows/issues/124)

**Review by pconrad — CHANGES_REQUESTED** [review]
Code LGTM!  No problems with the code.

But a few other concerns before we merge this...

It looks like you did follow up on repurposing this to a "suspend user" button.  However, I wonder if you redeployed on dokku?

The dokku deployment still says "Edit Commons".

Also, can you update the PR title and description to match the new focus?
Thanks!

**Comment by mczimmermann** [comment]
Just updated the dokku deployment and PR info!

**Review by nadavital — DISMISSED** [review]
LGTM

**Review by nadavital — APPROVED** [review]
LGTM

**Review by pconrad — APPROVED** [review]
LGTM

**Pull request merged**